msgpack/src/main/scala/fs2/data/msgpack/low/internal/ItemSerializer.scala

```diff
@@ -0,0 +1,247 @@
+/*
+ * Copyright 2024 fs2-data Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fs2
+package data
+package msgpack
+package low
+package internal
+
+import scodec.bits._
+
+private[low] object ItemSerializer {
+  class MalformedItemError extends Error("item exceeds the maximum size of it's format")
+  class MalformedStringError extends MalformedItemError
+  class MalformedBinError extends MalformedItemError
+  class MalformedIntError extends MalformedItemError
+  class MalformedUintError extends MalformedItemError
+
+  /** Checks whether integer `x` fits in `n` bytes. */
+  @inline
+  private def fitsIn(x: Int, n: Long): Boolean =
+    java.lang.Integer.compareUnsigned(x, (Math.pow(2, n.toDouble).toLong - 1).toInt) <= 0
+
+  private case class SerializationContext[F[_]](out: Out[F],
+                                                chunk: Chunk[MsgpackItem],
+                                                idx: Int,
+                                                rest: Stream[F, MsgpackItem])
+
+  /** Buffers [[Chunk]] into 4KiB segments before calling [[Pull.output]].
+    *
+    * @param contents buffered [[Chunk]]
+    */
+  private class Out[F[_]](contents: Chunk[Byte]) {
```

Comment:
I like the approach, however, this creates a new `Out` instance for each incoming item. If you run the benchmark with allocation and CPU flame graphs, does it appear to be a bottleneck?

Reply:
It doesn't seem like a significant bottleneck to me.

Here's a screenshot from VisualVM memory profiler:
![image](https://github.com/user-attachments/assets/e8f8ea70-e1c8-4267-a9a8-97123938fd92)

I cannot find any reference to `Out`'s methods in CPU samples so either I'm doing something wrong or these methods get optimized out :)

Reply:
I was thinking about runnin the benchmark with async-profiler outputting flamegraph. Something like:

```shell
$ sbt
> project benchmarksJVM
> Jmh/run fs2.data.benchmarks.MsgPackItemParserBenchmarks -prof "async:libPath=/path/to/libasyncProfiler.so;output=flamegraph;event=cpu;alluser=true"
```

Reply:
It takes around 10% of CPU and 17% of allocation graphs. 

Making the`Out` class an implicit around `Chunk[Byte]` lowers the results into 7% and 15% respectively. I'm not sure how could we optimize this more.

![image](https://github.com/user-attachments/assets/438f1454-f35d-4eb0-a140-22766bfaa597)
![image](https://github.com/user-attachments/assets/d61b6e27-305d-4d6b-a8e9-27974f6238a5)

Reply:
Time and memory seem to be dominated by `ByteVector` creation at every step. I need to review again these creations, and why they occur. But I could live with merging the current state and improving later on.